atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1077,6 +1077,34 @@ public void cleanup() {
         assertThat(counter.intValue()).isEqualTo(2);
     }
 
+    @Test
+    public void conditionCleanupRunsBeforeOnSuccess() {
+        MutableLong counter = new MutableLong(0L);
+        PreCommitCondition latchingCondition = new PreCommitCondition() {
+            @Override
+            public void throwIfConditionInvalid(long timestamp) {}
+
+            @Override
+            public void cleanup() {
+                assertThat(counter.getAndAdd(1)).isEqualTo(0);
+            }
+        };
+
+        serializableTxManager.runTaskWithConditionThrowOnConflict(latchingCondition, (tx, condition) -> {
+            tx.onSuccess(() -> assertThat(counter.getAndAdd(1)).isEqualTo(1));
+            tx.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("value")));
+            return null;
+        });
+        assertThat(counter.intValue()).isEqualTo(2);
+
+        counter.setValue(0);
+        serializableTxManager.runTaskWithConditionReadOnly(latchingCondition, (tx, condition) -> {
+            tx.onSuccess(() -> assertThat(counter.getAndAdd(1)).isEqualTo(1));
+            return tx.get(TABLE, ImmutableSet.of(TEST_CELL));
+        });
+        assertThat(counter.intValue()).isEqualTo(2);
```

Comment:
This is compact, but would suggest for readability to have a separate test for the rTWCTOC and rTWCRO cases, and ideally use a separate counter/condition in both; the reset halfway was a bit confusing